plaidml/edsl/tests/edsl_test.cc

```diff
@@ -1957,6 +1957,73 @@ TEST_F(CppEdsl, Lens) {
   checkExact(program, {input}, {expected});
 }
 
+TEST_F(CppEdsl, Loop) {
+  auto A = Placeholder(DType::FLOAT32, {4});
+  int64_t lb = 0;
+  int64_t hb = 5;
+  int64_t step = 1;
+  Tensor O = loop(/*lowerBound*/ lb, /*upperBound*/ hb, /*step*/ step,  //
+                  /*loop-carried variable*/ {A},                        //
+                  [&]() { return A + 1; });
+  auto program = makeProgram("loop", {A}, {O});
+  std::vector<float> input = {
+      1, 1, 1, 1  //
+  };
+  std::vector<float> expected = {
+      6, 6, 6, 6  //
+  };
+  checkExact(program, {input}, {expected});
+}
+
+TEST_F(CppEdsl, LoopConstantBuffer) {
+  auto A = Placeholder(DType::FLOAT32, {4});
+  size_t lb = 0;
+  size_t hb = 5;
+  size_t step = 1;
+  Tensor O = loop(/*lowerBound*/ lb, /*upperBound*/ hb, /*step*/ step,  //
+                  /*loop-carried variable*/ {A},                        //
+                  [&]() {
+                    std::vector<float> test{1, 1, 1, 1};
+                    auto B = Constant(makeBuffer(DType::FLOAT32, {4}, test), "test");
+                    return A + B;
+                  });
+  auto program = makeProgram("loop", {A}, {O});
+  std::vector<float> input = {
+      1, 1, 1, 1  //
+  };
+  std::vector<float> expected = {
+      6, 6, 6, 6  //
+  };
+  checkExact(program, {input}, {expected});
+}
+
+TEST_F(CppEdsl, LoopMultiIter) {
+  auto A = Placeholder(DType::FLOAT32, {4});
+  auto B = Placeholder(DType::FLOAT32, {4});
+  int64_t lb = 0;
+  int64_t hb = 5;
+  int64_t step = 1;
+  auto fn = [&]() -> TensorVec {
+    auto O = A + 1;
+    auto T = B + 1;
+    return {O, T};
+  };
+  auto output = loop(/*lowerBound*/ lb, /*upperBound*/ hb, /*step*/ step,  //
```

Comment:
Could you please add some calculations before the loop? like `A=A+3`, to double check the statements before loops are not included in the loop body?

Reply:
ok, no problem, actually, I wonder to add some more complex test case, I have made a small 
draft sequence case, according to liyang's representation about `RNNsequence`. but I run into some type conversion error, I'm still trying to fix it.